common/src/stack/command/stack/commands/set/host/power/__init__.py

```diff
@@ -8,96 +8,125 @@
-	command uses IPMI to change the power setting on a host.
-	
+	command uses IPMI for hardware based hosts to change the power setting.
+
 	<arg type='string' name='host' repeat='1'>
 	One or more host names.
 	</arg>
 
 	<param type='string' name='command' optional='0'>
-	The power command to execute. Valid power commands are: "on", "off" and "reset".
+	The power command to execute. Valid power commands are: "on", "off", "reset", and "status".
 	</param>
 
 	<param type='boolean' name='debug' optional='1'>
-	Print debug output from the ipmitool command.
+	Print debug output from the command. For ipmi capable hosts, prints
+	the output from ipmitool.
+	</param>
+
+	<param type='string' name='method' optional='1'>
+	Set a desired method for communicating to hosts, possible methods
+	include but are not limited to ipmi and ssh.
 	</param>
 
 	<example cmd='set host power stacki-1-1 command=reset'>
 	Performs a hard power reset on host stacki-1-1.
 	</example>
-	"""
-
-	def doPower(self, host, ipmi_ip, cmd):
-		import subprocess
-		import shlex
-
-		if not ipmi_ip:
-			return
-
-		username = self.getHostAttr(host, 'ipmi_username')
-		if not username:
-			username = 'root'
 
-		password = self.getHostAttr(host, 'ipmi_password')
-		if not password:
-			password = 'admin'
-
-		ipmi = 'ipmitool -I lanplus -H %s -U %s -P %s chassis power %s' \
-			% (ipmi_ip, username, password, cmd)
-
-		p = subprocess.Popen(shlex.split(ipmi), stdout = subprocess.PIPE,
-			stderr = subprocess.STDOUT)
-		out, err = p.communicate()
-
-		if self.debug:
-			self.beginOutput()
-			self.addOutput(host, out.decode())
-			self.endOutput(padChar='', trimOwner=True)
+	<example cmd='set host power stacki-1-1 command=off method=ssh'>
+	Turns off host stacki-1-1 using ssh.
+	</example>
+	"""
+	def mq_publish(self, host, cmd):
+		"""
+		Publish the power status to the
+		message queue
+		"""
 
 		ttl = 60*10
 		if cmd == 'off':
 			ttl = -1
 
-		msg = { 'source' : host, 
-			'channel': 'health', 
+		msg = { 'source' : host,
+			'channel': 'health',
 			'ttl'    : ttl,
 			'payload': '{"state": "power %s"}' % cmd }
 
 		tx = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-		tx.sendto(json.dumps(msg).encode(), 
+		tx.sendto(json.dumps(msg).encode(),
 			  ('127.0.0.1', stack.mq.ports.publish))
 		tx.close()
 
-
-
 	def run(self, params, args):
 		if not len(args):
 			raise ArgRequired(self, 'host')
 
-		cmd, debug = self.fillParams([ ('command', None, True), ('debug', False) ])
-
-		if cmd == 'status':
-			#
-			# used by "stack list host power" -- this is an easy way in which to
-			# share code between the two commands
-			#
-			# set 'debug' to True in order to get output from the status command
-			#
-			debug = True
-		elif cmd not in [ 'on', 'off', 'reset' ]:
-			raise ParamError(self, 'command', 'must be "on", "off" or "reset"')
-		
+		cmd, debug, force_imp = self.fillParams([
+			('command', None, True),
+			('debug', False),
+			('method', None)
+		])
+		imp_names = []
```

Comment:
Consider removing this and setting it in the `if: ... else: ...` statements below:
```
if force_imp and force_imp in self.impl_list:
    imp_names = [force_imp]
else:
    imp_names = ...
```

Since Python does not have block level scoping, these variables will be available outside the `if: ... else: ...` block.